pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -245,9 +314,137 @@ func (r *ReconcileNodeMaintenance) stopNodeMaintenance(nodeName string) error {
 		return err
 	}
 
+	if r.isLeaseSupported {
+		log.Info("Lease object supported, deleting lease")
+		if err := r.deleteLease(nodeName); err != nil {
+			return err
+		}
+	}
+
 	return nil
 }
 
+func (r *ReconcileNodeMaintenance) deleteLease(nodeName string) error {
+
+		nName := apitypes.NamespacedName{Namespace: LeaseNamespace, Name: nodeName}
+		lease := &coordv1beta1.Lease{}
+
+		if err := r.client.Get(context.TODO(), nName, lease); err != nil {
+			return err
+		}
+		r.client.Delete(context.TODO(), lease);
+		return nil
+}
+
+func (r *ReconcileNodeMaintenance) obtainLease(node *corev1.Node, durationInSeconds int) (bool, reconcile.Result) {
+
+	lease, needUpdate, err := r.createLease(node, durationInSeconds)
+
+	if err != nil {
+		log.Errorf("MaintOn: failed to create lease error=%v", err)
+		return false, reconcile.Result{Requeue: true, RequeueAfter: RequeWaitTime}
+	}
+
+	if needUpdate == false {
+		return true, reconcile.Result{}
+	}
+
+	log.Info("MaintOn: update lease - lease duration change requested")
+
+	if _, err = r.updateLease(node, lease, durationInSeconds, TransitionNoChange, false); err != nil {
+		return false, reconcile.Result{Requeue: true, RequeueAfter: RequeWaitTime}
+	}
+
+	return true, reconcile.Result{}
+}
+
+func (r *ReconcileNodeMaintenance) createLease(node *corev1.Node, durationInSeconds int) (*coordv1beta1.Lease, bool, error) {
+	nodeName := node.ObjectMeta.Name
+
+	holderIdentity := LeaseHolderIdentity
+
+	owner := &metav1.OwnerReference{
+		APIVersion: "v1",
+		Kind:       "Node",
+		Name:       nodeName,
+		UID:        node.ObjectMeta.UID,
+	}
+	leaseTransitions := int32(1)
+	microTimeNow := metav1.NewMicroTime(time.Now())
+
+	tmpDurationInSeconds := int32(durationInSeconds)
+	var refDuration *int32 =  &tmpDurationInSeconds
+
+	lease := &coordv1beta1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            nodeName,
+			Namespace:       LeaseNamespace,
+			OwnerReferences: []metav1.OwnerReference{*owner},
+		},
+		Spec: coordv1beta1.LeaseSpec{
+			HolderIdentity:       &holderIdentity,
+			LeaseDurationSeconds: refDuration,
+			AcquireTime:          &microTimeNow,
+			RenewTime:            &microTimeNow,
+			LeaseTransitions:     &leaseTransitions,
+		},
+	}
+
+	if err := r.client.Create(context.TODO(), lease); err != nil {
+		if errors.IsAlreadyExists(err) {
+			return nil, true, nil
+		}
+		return nil, false, err
+	}
+	return lease, false,  nil
+}
+
+func (r *ReconcileNodeMaintenance) updateLease(node *corev1.Node, lease *coordv1beta1.Lease, durationInSeconds int, transitions TransitionAction, resetDuration bool) (*coordv1beta1.Lease, error) {
```

Comment:
```suggestion
func (r *ReconcileNodeMaintenance) updateLease(node *corev1.Node, lease *coordv1beta1.Lease) (*coordv1beta1.Lease, error) {
```

Reply:
`updateLease()` is called from only one place, meaning that `durationInSeconds`, `transitions`, and `resetDuration` are all fixed. 

You need to calculate these based on the existing values of the lease.

Reply:
Currently the lease period is a constant value, but you may want to change this in a possible future change the CRD, therefore I am passing these values a as arguments.

Reply:
Ok, but the same does not apply to `transitions`, and `resetDuration`.
You *must* calculate them inside the `updateLease()` function based on the existing contents of the lease and how they compare to the fields you need to update. 

It would be worthwhile restructuring this function so that you can more easily write unit tests that verify that `.Spec.LeaseTransitions` and `.Spec.AcquireTime` are updated correctly.  Currently both are almost always set incorrectly.